Add metadynamics

vitartas opened the pull request:
Integrated PLUMED to mlp-train via ASE calculator. Still need to include Metadynamics methods for data analysis.

tristan-j-wood commented on mlptrain/log.py:
```diff
@@ -10,9 +10,6 @@
 # Try and use colourful logs
 try:
     import coloredlogs
-    FORMAT = '%(asctime)s %(hostname)s %(name)s[%(processName)s] ' \
```
What was the reason for removing this again?

vitartas replied:
I've decided to return to using PIDs in log as I've used PIDs in naming PLUMED output files when multiple metadynamics runs execute in parallel.

tristan-j-wood commented on mlptrain/sampling/md.py:
```diff
@@ -24,7 +26,8 @@ def run_mlp_md(configuration: 'mlptrain.Configuration',
-               bias:          Optional['mlptrain.bias.Bias'] = None,
+               bias:          Optional[Union['mlptrain.Bias',
+                                             'mlptrain.PlumedBias']] = None,
                method:        Optional[str] = None,
```
I would only allow valid options. At the moment people can type anything in and it won't immediately break if it isn't implemented, which might lead to confusion over what is breaking. Also for strings, lower() so people can type in any case

tristan-j-wood commented on mlptrain/sampling/md.py:
```diff
@@ -99,17 +105,39 @@ def append_energy(_atoms=ase_atoms):
-    dyn.run(steps=n_steps)
+
+    if isinstance(bias, PlumedBias):
+        from ase.calculators.plumed import Plumed
```
Can add some logs saying it is running plumed

tristan-j-wood commented on mlptrain/sampling/md.py:
```diff
@@ -99,17 +105,39 @@ def append_energy(_atoms=ase_atoms):
+        plumed_calc.plumed.finalize()
+
+    else:
+        ase_atoms.calc = mlp.ase_calculator
```
Logs saying it is running ase

tristan-j-wood commented on mlptrain/sampling/md.py:
```diff
@@ -227,3 +255,43 @@ def _n_simulation_steps(dt: float,
+                       f'FILE=HILLS_pid{os.getpid()}.dat']
+        setup.extend(metad_setup)
+
+    # Print
```
Slightly more informative comment might be useful

tristan-j-wood commented on mlptrain/sampling/metadynamics.py:
```diff
@@ -0,0 +1,165 @@
+                                   height=height,
+                                   biasfactor=biasfactor)
+
+        metad_processes, metad_trajs = ([] for _ in range(2))
```
I generally don't use this for small numbers of lists. It's longer and more difficult to read than simply: = [], []

tristan-j-wood commented on mlptrain/sampling/metadynamics.py:
```diff
@@ -0,0 +1,165 @@
+                metad_trajs.append(metad_process.get())
+
+        if save_sep:
+            os.mkdir('metad_trajectories')
```
What happens if the directory exists?

tristan-j-wood commented on mlptrain/sampling/plumed.py:
```diff
@@ -0,0 +1,285 @@
+                 atom_groups: Sequence = None,
+                 file_name: str = None):
+        """
+        This class contains methods to initialise PLUMED collective variables
```
If it shouldn't be called directly, would it be better to make it a private class?

tristan-j-wood commented on mlptrain/sampling/plumed.py:
```diff
@@ -0,0 +1,285 @@
+
+        for name, action in _names_and_actions:
+            if action != 'GROUP':
+                self.dof_names.append(f'{self.name}_{name}')
```
Need to return None

tristan-j-wood reviewed (commented):
Very nice code

vitartas commented:
Adding new commits:
1. Attempted to fix the code based on your comments
2. Contracted the method variable in run_mlp_md() to mlptrain.Bias and mlptrain.PlumedBias as the variable is not supposed to be changed by the user
3. Added a feature to fully initialise mlptrain.PlumedBias using a full PLUMED input file (which allows to execute run_mlp_md with an arbitrary PLUMED input)
4. Added an option to specify the component of the collective variable to use as the actual CV when initialised using mlptrain.PlumedCustomCV

tristan-j-wood reviewed (commented):
Few cosmetic changes

vitartas commented:
Adding new commits:
1. Refactored run_mlp_md()
2. Added a restart option for run_mlp_md() and run_metadynamics()
3. Added methods estimate_width() and try_multiple_biasfactors() which are helpful for choosing metadynamics parameters
4. Added plot_fes_convergence() which is useful for checking whether the FES has converged
5. Added compute_fes() and plot_fes()

tristan-j-wood commented on mlptrain/sampling/md.py:
```diff
@@ -1,42 +1,48 @@
+               bias:               Optional = None,
+               restart_files:      Optional[List[str]] = None,
+               copied_substrings:  Sequence[str] = ('.xml', '.json', '.pth'),
+               kept_substrings:    Sequence[str] = None,
```
Add arg for plumed print setup with docstring = False

tristan-j-wood commented on mlptrain/sampling/metadynamics.py:
```diff
@@ -341,41 +325,64 @@ def run_metadynamics(self,
-        if save_sep:
-            for idx, metad_traj in enumerate(metad_trajs):
-                metad_traj.save(filename=f'metad_{idx+1}.xyz')
+        self._move_and_save_files(metad_trajs, save_sep, all_to_xyz, restart)
```
Could have a comment explaining where files are being moved

tristan-j-wood commented on mlptrain/sampling/plumed.py:
```diff
@@ -434,12 +437,13 @@ def __init__(self,
-            label:       Optional[str] = None
-            ) -> None:
+def plot_cv_versus_time(filename:    str,
+                        style:       str = 'scatter',
```
Default could/_should_ be plot

tristan-j-wood commented on mlptrain/sampling/plumed.py:
```diff
@@ -215,6 +216,68 @@ def dof_sequence(self) -> str:
+
+            location: (float) Value of the CV where the wall will be located
+
+            kappa: (float) The force constant of the wall
```
Add units

tristan-j-wood commented on mlptrain/sampling/plumed.py:
```diff
@@ -215,6 +216,68 @@ def dof_sequence(self) -> str:
+
+            location: (float) Value of the CV where the wall will be located
+
+            kappa: (float) The force constant of the wall
```
Add units

tristan-j-wood commented on mlptrain/sampling/metadynamics.py:
```diff
@@ -715,43 +723,83 @@ def block_analysis(self,
         -----------------------------------------------------------------------
         Arguments:
 
+            energy_units: (str) Energy units to be used in plotting
```
Could specify the allowed units
